{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bb920f53",
   "metadata": {},
   "source": [
    "### Open Reading Frame; Biopython\n",
    "\n",
    "- Biopython does not have one-line `.find-orfs()` method \n",
    "- This notebook is dedicated to manually building ORF from scratch\n",
    "\n",
    "***\n",
    "\n",
    "\n",
    "\n",
    "> A continuous stretch of codons without stop codons, beginning with a start codon (usually ATG) and ending at a stop (TAA, TAG, or TGA).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c76c7bc5",
   "metadata": {},
   "source": [
    "### 1. Get the `genomic sequence`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09fde74b",
   "metadata": {},
   "source": [
    "**Escherichia coli strain PNUSAE211155, whole genome shotgun sequencing project**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d8a6257",
   "metadata": {},
   "source": [
    "Escherichia coli strain PNUSAE211155, whole genome shotgun sequencing project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d4e7779",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez,SeqIO,SeqUtils, SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b41f8e38",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = 'jasovicluka1@gmail.com'\n",
    "handle = Entrez.esearch(db='nucleotide',term='E. coli', retmax=100,idtype='acc')\n",
    "record = Entrez.read(handle,'gb')\n",
    "handle.close()\n",
    "\n",
    "id_list = record['IdList']\n",
    "id_list\n",
    "selected_ids = []\n",
    "for i,id in enumerate(id_list):\n",
    "    selected_ids = id_list[17]\n",
    "print(f'My selected ID: {selected_ids}')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8276dad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fetch = Entrez.efetch(db='nucleotide',id=selected_ids,rettype='gb',retmode='text')\n",
    "read = SeqIO.read(fetch,'gb')\n",
    "handle.close()\n",
    "record_1 = read.seq\n",
    "print(len(record_1))\n",
    "\n",
    "\n",
    "#Save the complete file\n",
    "\n",
    "with open('ORF_1.txt', 'w') as file_01:\n",
    "    SeqIO.write(read,file_01,'fasta')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2730fa7",
   "metadata": {},
   "source": [
    "### 2. ORF "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25bc3ec3",
   "metadata": {},
   "source": [
    "In bacteria, the most common start codon is `AUG` which codes for N-formylmethionine (**fMet**)\n",
    "\n",
    "Apart from this start codon, there are some alternative start codons, and the most common ones are:\n",
    "\n",
    "- `GUG`\n",
    "\n",
    "- `UUG`\n",
    "\n",
    "These start codons are recognized by the initiator tRNA carrying fMet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18eb6726",
   "metadata": {},
   "outputs": [],
   "source": [
    "# AUG - ATG \n",
    "# GUG - GTG\n",
    "# UUG - TTG "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6664866",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Full Sequence lenght: 4073392\n",
    "print(f'Full sequence lenght is {len(record_1)}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a9ab66c",
   "metadata": {},
   "source": [
    "#### 2.1 Load the full sequence, fasta format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46e5717c",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('ORF_1.txt','r') as file:\n",
    "    record = SeqIO.read(file,\"fasta\")\n",
    "sequence = record.seq\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655c7d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc9757c",
   "metadata": {},
   "source": [
    "#### 2.2 Iterate over each of the 6 possible reading frames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a243c8f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#START:ATG STOP:TAA, TGA TAG\n",
    "\n",
    "reverse_complement = sequence.reverse_complement()\n",
    "reverse_complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befd9d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "seq_str = str(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4903a6bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad3a21e2",
   "metadata": {},
   "source": [
    "Reading frame 0 → starts at position 0:\n",
    "ATG, CAA, TGA, ...\n",
    "\n",
    "Reading frame 1 → starts at position 1:\n",
    "TGC, AAT, GAT, ...\n",
    "\n",
    "Reading frame 2 → starts at position 2:\n",
    "GCA, ATG, ATA, ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e3d035",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loop through the complete sequence\n",
    "# start storing when hit the start codon\n",
    "# stop when hit the stop codon\n",
    "\"\"\"\n",
    "#ORF 0\n",
    "import re\n",
    "#rf1 = []\n",
    "#rf2 = []\n",
    "#rf3 = []\n",
    "rfs = [[],[],[]]\n",
    "#Checkpoint #1 - Searching for start codon \"ATG\" and it's position in the full sequence\n",
    "for i in [0,1,2]:\n",
    "    found_item = False\n",
    "    for j in range(i,len(sequence)-2,3):\n",
    "        current_frame = []\n",
    "        codons = seq_str[j:j+3]\n",
    "        #print(codons)\n",
    "        if codons == 'ATG' and not found_item:\n",
    "            print(f'Found start of the current frame, ORF {i}')\n",
    "            found_item = True\n",
    "        if found_item:\n",
    "            current_frame.append(codons)\n",
    "            print(current_frame)\n",
    "            if codons in ['TAA','TGA','TAG']:\n",
    "                found_item = False\n",
    "                rfs[i].append(current_frame)\n",
    "\n",
    "print(len(rfs[0]))\n",
    "print(len(rfs[1]))\n",
    "print(len(rfs[2]))\n",
    "\n",
    "\n",
    "\"\"\"   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4dbb6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def orf_reader(sequence,found_item=True):\n",
    "    rfs1 =[[],[],[]]\n",
    "    for i in [0,1,2]:\n",
    "        found_item= False\n",
    "        current_frame = []\n",
    "        for j in range(i,len(sequence) -2,3):\n",
    "            codons = str(sequence[j:j+3])                        \n",
    "            if codons == 'ATG' and not found_item:\n",
    "                current_frame =['ATG']\n",
    "                found_item = True\n",
    "                continue # to escape adding 2x ATG\n",
    "            if found_item:\n",
    "                current_frame.append(codons)\n",
    "                if codons in ['TAA','TAG','TGA']:\n",
    "                    rfs1[i].append(current_frame)\n",
    "                    current_frame = []\n",
    "                    found_item = False\n",
    "\n",
    "    \n",
    "    return rfs1\n",
    "rfs0 = orf_reader(sequence)\n",
    "print(rfs0)\n",
    "\n",
    "#orf_reader(sequence)\n",
    "\n",
    "# Sta hocu da uraidm\n",
    "# funckija koja ce da procita sve sto je storovano iz sequence , i reverse complement "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a427e7b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_0 = rfs0[0]\n",
    "orf_1 = rfs0[1]\n",
    "orf_2 = rfs0[2]\n",
    "print(len(orf_0))\n",
    "print(len(orf_1))\n",
    "print(len(orf_2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cad9dae",
   "metadata": {},
   "source": [
    "### Control point below:\n",
    "\n",
    "- Testing `orf_reader ()` if it works on synthetic sequence before I continue to next part \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9dc5915",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "synthetic_seq = Seq(\"ATGAAATAATATGAGATAAATGAAGGTAG\")\n",
    "orf_reader(synthetic_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcc4131f",
   "metadata": {},
   "outputs": [],
   "source": [
    "synth_reverse = synthetic_seq.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "adcd4600",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[['ATG', 'AAA', 'TAA']], [['ATG', 'AGA', 'TAA']], []]\n"
     ]
    }
   ],
   "source": [
    "reslt = orf_reader(synthetic_seq,found_item=True)\n",
    "print(reslt)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b66c808",
   "metadata": {},
   "source": [
    "#### Wrapper function that will get all the readingframes (`forward` + `reverse` strand -> 6 ORFS total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5256c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Control point, results passed from the previous func.: [[['ATG', 'AAA', 'TAA']], [['ATG', 'AGA', 'TAA']], []]\n",
      "2nd Control pont, synthetic reverse strand ORFs: [[], [], []]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{0: 345}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def scan_all_frames(seq1,seq2):\n",
    "    all_orfs = {}\n",
    "    reslt = orf_reader(seq1,found_item=True)\n",
    "    print(f'Control point, results passed from the previous func.: {reslt}')\n",
    "    reslt_rev = orf_reader(seq2,found_item=True)\n",
    "    print(f\"2nd Control pont, synthetic reverse strand ORFs: {reslt_rev}\")\n",
    "    key1 = {}\n",
    "    key =0\n",
    "    val = {}\n",
    "    for x in range(len(reslt)):\n",
    "        for j in (reslt[x]):\n",
    "            \n",
    "        return key\n",
    "scan_all_frames(synthetic_seq,synth_reverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f439f1",
   "metadata": {},
   "source": [
    "moras da proveris ponasanje i kako bi najprostije odradio sto ti padne na pamet"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
